### PR TITLE
scarab: Apply scaling factor in Wayland

### PR DESCRIPTION
**Pull request by l0b0**
Sets `AVALONIA_GLOBAL_SCALE_FACTOR` to the GNOME desktop scaling factor based on
<https://github.com/AvaloniaUI/Avalonia/issues/9390#issuecomment-2382126451>.

Also:

- Removed seemingly unnecessary dependencies
- Formatted using `shfmt --write pkgs/tools/games/scarab/scaling-settings.bash` (necessitated copying the relevant `.editorconfig` settings)
- Linted using `shellcheck --enable=all --severity=style pkgs/tools/games/scarab/scaling-settings.bash`

Thanks to @Artturin for help on Matrix!

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] `shellcheck`ed
- [x] `shfmt`ed
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/tools/games/scarab/default.nix [hunk 1]
```diff
@@ -2,12 +2,15 @@
...
   fetchFromGitHub,
+  bc,
+  glib,
   glibc,
   zlib,
   gtk3,
   copyDesktopItems,
   icoutils,
   wrapGAppsHook3,
+  writeText,
```
Comment (huantianad) [inline pkgs/tools/games/scarab/default.nix:13]: leftover input

**Review by huantianad — APPROVED** [review]
Looks pretty good, I am curious though, since this is an avalonia issue, anyway we can apply this for all avalonia apps?

**Comment by l0b0** [comment]
> Looks pretty good, I am curious though, since this is an avalonia issue, anyway we can apply this for all avalonia apps?

Good question. I'm not at all familiar with Avalonia, and a quick search for it in nixpkgs didn't reveal anything obvious. @corngood any ideas?

**Comment by corngood** [comment]
`pkgs/build-support/dotnet/fetch-nupkg/overrides.nix` has:

```
  "Avalonia.X11" =
    package:
    package.overrideAttrs (
      old:
      lib.optionalAttrs (!stdenv.hostPlatform.isDarwin) {
        setupHook = writeText "setupHook.sh" ''
          prependToVar dotnetRuntimeDeps \
            "${lib.getLib libICE}" \
            "${lib.getLib libSM}" \
            "${lib.getLib libX11}"
        '';
      }
    );
```

This is probably where you'd want to do it.  You probably couldn't use `dotnetRuntimeDeps`, but you might be able to use `makeWrapperArgs`.  I think something like `--run "source ""${./scaling-settings.bash}"` might allow you to modify the environment in the wrapper, but I haven't tried it.

**Comment by huantianad** [comment]
Wanted to bump this again, do we want to apply this change just for scarab for now, and for all avalonia apps later?

**Comment by corngood** [comment]
From the upstream issue:

> Avalonia [currently reads Xft.dpi value from XRDB](https://github.com/AvaloniaUI/Avalonia/blob/7a7bee4cb4fa054d6780acb3b9138b822c3152dc/src/Avalonia.X11/Screens/X11Screens.Scaling.cs#L58-L61), which is the only thing we can do since WM/DE authors simply refuse to provide accurate per-monitor scaling information for X11 apps.

It seems like we're reproducing this logic.  I don't understand why it's not working out of the box.

**Comment by l0b0** [comment]
> It seems like we're reproducing this logic. I don't understand why it's not working out of the box.

Same, but this works right now. I'd rather just revert once a more suitable solution is found.

File: pkgs/tools/games/scarab/scaling-settings.bash [hunk 2]
```diff
@@ -0,0 +1,17 @@
+# Keep existing value if it is already non-empty
+if [[ -z "${AVALONIA_GLOBAL_SCALE_FACTOR-}" ]]; then
+    # Attempt to get GNOME desktop interface scaling factor
+    AVALONIA_GLOBAL_SCALE_FACTOR="$(gsettings get org.gnome.desktop.interface scaling-factor | cut --delimiter=' ' --fields=2)"
+fi
+
+if [[ -z "${AVALONIA_GLOBAL_SCALE_FACTOR-}" ]]; then
+    # Attempt to get scaling factor from X FreeType DPI setting
+    dpi="$(xrdb -get Xft.dpi)"
```
Comment (corngood) [inline pkgs/tools/games/scarab/scaling-settings.bash:9]: Should we check for existence of `xrdb`?

File: pkgs/tools/games/scarab/scaling-settings.bash [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    if [[ -n "$dpi" ]]; then
+        AVALONIA_GLOBAL_SCALE_FACTOR=$(echo "scale=2; $dpi/96" | bc)
```
Comment (corngood) [inline pkgs/tools/games/scarab/scaling-settings.bash:11]: Should we use bash arithmetic here to avoid the `bc` dependency?

**Comment by huantianad** [comment]
> From the upstream issue:
> 
> 
> 
> > Avalonia [currently reads Xft.dpi value from XRDB](https://github.com/AvaloniaUI/Avalonia/blob/7a7bee4cb4fa054d6780acb3b9138b822c3152dc/src/Avalonia.X11/Screens/X11Screens.Scaling.cs#L58-L61), which is the only thing we can do since WM/DE authors simply refuse to provide accurate per-monitor scaling information for X11 apps.
> 
> 
> 
> It seems like we're reproducing this logic.  I don't understand why it's not working out of the box.
> 
> 
> 
> 

The issue is really strange, it does sound like this should be things that upstream should be and already is doing. Perhaps it has to do with XWayland, and how different DEs handle scaling with X11 apps running in XWayland.

File: pkgs/tools/games/scarab/scaling-settings.bash (hunk 3 above)
Comment (l0b0) [reply pkgs/tools/games/scarab/scaling-settings.bash:11]: Bash only does integer arithmetic, so it would break in case of non-integer scaling. We should probably support non-integer scaling if possible.

Comment (corngood) [reply pkgs/tools/games/scarab/scaling-settings.bash:11]: Ah sorry, I'm so used to doing it in zsh that I forgot it wasn't standard.

File: pkgs/tools/games/scarab/scaling-settings.bash [hunk 4]
```diff
@@ -0,0 +1,17 @@
+# Keep existing value if it is already non-empty
+if [[ -z "${AVALONIA_GLOBAL_SCALE_FACTOR-}" ]]; then
+    # Attempt to get GNOME desktop interface scaling factor
+    AVALONIA_GLOBAL_SCALE_FACTOR="$(gsettings get org.gnome.desktop.interface scaling-factor | cut --delimiter=' ' --fields=2)"
```
[the remark this answers is not in the document]
Comment (l0b0) [reply pkgs/tools/games/scarab/scaling-settings.bash:4]: Since it doesn't make sense to run Scarab in a pure shell, I think you're right.

[the remark this answers is not in the document]
Comment (corngood) [reply pkgs/tools/games/scarab/scaling-settings.bash:4]: For the setting I get `uint32 0`, using xmonad on xorg.

`gsettings describe` says

> Integer factor used to scale windows by. For use on high-dpi screens. 0 means pick automatically based on monitor.

So I guess falling through to the other methods would make sense.

[the remark this answers is not in the document]
Comment (l0b0) [reply pkgs/tools/games/scarab/scaling-settings.bash]: Fixed by assuming that a zero from `gsettings` means I'd also get the same or another meaningless value from `xrdb`, and simply unsetting the scaling factor if that happens.

**Comment by l0b0** [comment]
All done @corngood, thank you!

File: pkgs/tools/games/scarab/default.nix [hunk 5]
```diff
@@ -1,13 +1,10 @@
 {
   lib,
+  bc,
   buildDotnetModule,
   fetchFromGitHub,
-  glibc,
```
Comment (corngood) [inline pkgs/tools/games/scarab/default.nix:5]: Removing these is good if they aren't needed, but I think it should be a separate commit from the wrapper, in case we need to bisect/revert.

File: pkgs/tools/games/scarab/scaling-settings.bash [hunk 6]
```diff
@@ -0,0 +1,24 @@
+# Keep existing value if it is already non-empty
+if [[ -z "${AVALONIA_GLOBAL_SCALE_FACTOR-}" ]] && type gsettings >/dev/null; then
+    echo 'Attempting to get GNOME desktop interface scaling factor'
```
Comment (corngood) [inline pkgs/tools/games/scarab/scaling-settings.bash:3]: I would probably have made these echos go to stderr, since this is a wrapper.

**Review by corngood — APPROVED** [review]
This all looks good to me.  Thanks!

The only question I have is: are there any automated editorconfig checks that could possibly fail because of the added coverage of .bash files?

I'll try to answer that myself, but if maybe someone knows for sure.

**Comment by l0b0** [comment]
> The only question I have is: are there any automated editorconfig checks that could possibly fail because of the added coverage of .bash files?

Oh, right! Should be alright, `editorconfig-checker $(find . -name '*.bash')` prints nothing and exits with 0.

**Comment by l0b0** [comment]
> The only question I have is: are there any automated editorconfig checks that could possibly fail because of the added coverage of .bash files?

Oh, good point! `editorconfig-checker $(find . -name '*.bash')` does find a handful of wrong indentations, that's all. I'll fix those, and move the change to a separate PR.

**Comment by corngood** [comment]
> are there any automated editorconfig checks that could possibly fail because of the added coverage of .bash files?

... There are editorconfig checks.  They are usually only run on files changed in the PR.  There are quite a few existing errors in master when it's run on the whole repo, but I verified that there are no errors in .bash files on this PR.

**Comment by corngood** [comment]
> I'll fix those in a separate PR. I've removed that commit.

@l0b0 I ran the same checks, but with `-disable-indent-size`, because that's what seems to be used in the workflow:

https://github.com/NixOS/nixpkgs/actions/runs/11760835386/workflow?pr=348427

So if it's not going to fail in the regular checks, I think it's fine if you don't fix the other files.  It wouldn't hurt though.  Up to you...

**Comment by l0b0** [comment]
> > I'll fix those in a separate PR. I've removed that commit.
> 
> @l0b0 I ran the same checks, but with `-disable-indent-size`, because that's what seems to be used in the workflow:
> 
> [NixOS/nixpkgs/actions/runs/11760835386/workflow?pr=348427](https://github.com/NixOS/nixpkgs/actions/runs/11760835386/workflow?pr=348427)
> 
> So if it's not going to fail in the regular checks, I think it's fine if you don't fix the other files. It wouldn't hurt though. Up to you...

OIC, thanks! I didn't look at the indent size check. I'll leave it in for now, and see what people say. I assume the indent check is only disabled temporarily, until more files pass, but worst case that just means we'll commit the .editorconfig change itself.

**Comment by corngood** [comment]
Sorry, I should have actually tried to run this before.  Without `gsettings` available I get this log noise:

```
/nix/store/z3fbayxg8aj5acjpmyfk6ln2cws8fd0l-scaling-settings.bash: line 2: type: gsettings: not found
```

I think using `command -v` instead of `type` would fix this.  Or maybe `type -p`?  I'm not 100% sure of the pros/cons of those.  Looks like the former is a bit more common in the repo.

If `$DISPLAY` is not set, I `xrdb` fails and doesn't try to start Scarab.  That's presumably the same with any failure of either `xrdb` or `gsettings`.  In the case of `DISPLAY` not being set, the app won't even run, so maybe it's not a big deal.  However, if we were going to apply this script to more packages, we'd have to be careful.  I believe some of them have non-graphical interfaces using the same executable (even if it's just something like `--version` handling).


Also...

On my laptop I'm using gnome+wayland, and 192 dpi.  For some reason, `gsettings` returns `uint32 0`, and `xrdb` reports `192`.  Because the check for 0 is after the `xrdb` block, it never actually tries to run `xrdb`, and launches without a scaling setting.  Perhaps the 0 check should be specific to `gsettings`, since it's a documented default there.

**Comment by l0b0** [comment]
> Sorry, I should have actually tried to run this before. Without `gsettings` available I get this log noise:
> 
> ```
> /nix/store/z3fbayxg8aj5acjpmyfk6ln2cws8fd0l-scaling-settings.bash: line 2: type: gsettings: not found
> ```
> 
> I think using `command -v` instead of `type` would fix this. Or maybe `type -p`? I'm not 100% sure of the pros/cons of those. Looks like the former is a bit more common in the repo.

Yes, `command -v` is usually recommended over `type`. I originally used `type` in the (admittedly extremely unlikely) case that someone was using a function or alias to point to a `gsettings` not on `$PATH`, but log noise is probably a good enough reason to use `command -v` instead.
 
> If `$DISPLAY` is not set, I `xrdb` fails and doesn't try to start Scarab. That's presumably the same with any failure of either `xrdb` or `gsettings`. In the case of `DISPLAY` not being set, the app won't even run, so maybe it's not a big deal. However, if we were going to apply this script to more packages, we'd have to be careful. I believe some of them have non-graphical interfaces using the same executable (even if it's just something like `--version` handling).

We probably won't use this for other packages. It's really just a one-off fix for a persistent issue with Scarab specifically. In the unlikely event that someone reuses this script I'd expect they'd make sure to deal with non-GUI situations if relevant.

> On my laptop I'm using gnome+wayland, and 192 dpi. For some reason, `gsettings` returns `uint32 0`, and `xrdb` reports `192`. Because the check for 0 is after the `xrdb` block, it never actually tries to run `xrdb`, and launches without a scaling setting. Perhaps the 0 check should be specific to `gsettings`, since it's a documented default there.

Thanks! I've moved the check.

**Review by corngood — APPROVED** [review]
